hydromt/components/datasets.py

```diff
@@ -0,0 +1,407 @@
+"""Xarrays component."""
+
+from pathlib import Path
+from shutil import move
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    List,
+    Optional,
+    Union,
+    cast,
+)
+
+from pandas import DataFrame
+from xarray import DataArray, Dataset
+
+from hydromt._typing.type_def import XArrayDict
+from hydromt.components.base import ModelComponent
+from hydromt.hydromt_step import hydromt_step
+from hydromt.io.readers import read_nc
+from hydromt.io.writers import write_nc
+
+if TYPE_CHECKING:
+    from hydromt.models.model import Model
+
+_DEFAULT_DATASET_FILENAME = "datasets/{name}.nc"
+
+
+class DatasetsComponent(ModelComponent):
+    """A component to manage colections of Xarray objects.
+
+    It contains a dictionary of xarray DataArray or DataSet objects.
+    """
+
+    def __init__(self, model: "Model", filename: str = _DEFAULT_DATASET_FILENAME):
+        """Initialize a Datasets.
+
+        Parameters
+        ----------
+        model: Model
+            HydroMT model instance
+        filename: str
+            The path to use for reading and writing of component data by default.
+            by default "datasets/{name}.nc" ie one file per dataset in the data dictionnary.
+        """
+        self._data: Optional[XArrayDict] = None
+        self._filename = filename
+        self._tmp_data_dir = None
+        self._defered_file_closes = []
+        super().__init__(model=model)
+
+    @property
+    def data(self) -> XArrayDict:
+        """Model data in the form of xarray objects.
+
+        Return dict of xarray.Dataset or xarray.DataArray objects
+        """
+        if self._data is None:
+            self._initialize()
+
+        assert self._data is not None
+        return self._data
+
+    def _initialize(self, skip_read=False) -> None:
+        """Initialize geoms."""
+        if self._data is None:
+            self._data = dict()
+            if self._root.is_reading_mode() and not skip_read:
+                self.read()
+
+    def set(
+        self,
+        data: Union[Dataset, DataArray],
```

Comment:
I guess the conversation was outdated but we can add DataFrame here and keep the lines that convert to_xarray